File: src/codility/StoneWall.java
package codility;

import java.util.Stack;

public class StoneWall {
    public int solution(int[] H) {

        //e.g. H[0] = 2 would be a block of size (0, 2)
        Block currentBlock = new Block(0, H[0]);
        Stack<Block> blockStack = new Stack();

        blockStack.add(currentBlock);
        int blocksRequired = 1;

        for (int i = 1; i < H.length; i++) {
            Block topStackBlock = blockStack.peek(); //remove any stack blocks that are taller than current block
            while (topStackBlock.upperHeight > H[i]) {
                blockStack.pop();
                if (!blockStack.isEmpty()) {
                    topStackBlock = blockStack.peek();
                } else break;
            }

            if (blockStack.isEmpty()) {
                blockStack.push(new Block(0, H[i]));
                blocksRequired++;
            }
            //block already exists in stack
            else if (blockStack.peek().upperHeight - H[i] == 0) {
                continue;
            }
            //put in a new block - need to calculate delta between tallest stack block and current block
            else {
                topStackBlock = blockStack.peek();
                blockStack.push(new Block(blockStack.peek().upperHeight, H[i]));
                blocksRequired++;
            }
        }

        return blocksRequired;
    }

    class Block {
        int lowerHeight;
        int upperHeight;

        Block(int pLowerHeight, int pUpperHeight) {
            lowerHeight = pLowerHeight;
            upperHeight = pUpperHeight;
        }
    }
}

File: src/codility/StoneWall.kt
package codility

import java.util.*

fun solutionwall(H: IntArray): Int {

    val stack = Stack<Int>()
    var count = 0

    if (H.size == 1) {
        return 1
    }

    stack.push(H[0])
    count = 1


    for (i in 1 until H.size) {
        if (stack.peek() != H[i]) {
            count += 1
        }

        stack.push(H[i])

    }

    return count
}

fun solutionwall2(H: IntArray): Int {
    val currentBlock = Block2(0, H[0])
    val blockStack = Stack<Block2>()

    blockStack.push(currentBlock)
    var blockRequired = 1
    for (i in 1 until H.size){
        var topStackBlock2 = blockStack.peek()
        while (topStackBlock2.higherHeight > H[i]){
            blockStack.pop()
            if (!blockStack.isEmpty()){
                topStackBlock2 = blockStack.peek()
            } else break
        }

        if (blockStack.isEmpty()){
            blockStack.push(Block2(0, H[i]))
            blockRequired++
        } else if(blockStack.peek().higherHeight - H[i] == 0){
            continue
        }
        else {
            topStackBlock2 = blockStack.peek()
            blockStack.push(Block2(blockStack.peek().higherHeight, H[i]))
            blockRequired++
        }
    }
    return blockRequired
}

data class Block2(val lowerHeight: Int, val higherHeight: Int)

fun main() {
    val scan = Scanner(System.`in`)

    val A = scan.nextLine().split(" ").map { it.trim().toInt() }.toIntArray()

    val solution = solutionwall(A)
    print("$solution ")
}
